### DISCUSSION TITLE
boolean attribute is ambiguous in lit element property.

### DISCUSSION THREAD

**Discussion by dbwodlf3** [lit/lit#2959] [General]
```ts
@customElement('example-component-container')
class ExampleComponentContainer extends LitElement{

  render(){
    return html`<example-component ?example=${false} />`
  }
}

@customElement('example-component')
class ExampleComponent extends LitElement{
  @property({type:Boolean})
  example=true

  render(){
    return html`<div></div>`
  }
  
  firstUpdated(){
    console.log(this.example);
  }
}
```

```html
<example-component-container />
```

```console
// True output, no false.
true
```

????

**Comment by justinfagnani** [comment]
The default is `true`, and since the parent hard-codes a boolean attribute to `false` the attribute will never be present and never removed, so the default will never be overridden.

It probably doesn't make much sense to have a `type: Boolean` property default to true. Maybe we should add a lit-analyzer lint rule for this.

**Reply by deanhaleem** [reply]
Hey @justinfagnani jw, if we do have a boolean prop defaulting to `true`, do you suggest using a `type: String` with a custom converter instead?

```ts
@property({
  type: String,
  attribute:'boolean-attribute',
  converter: (value) => {
    return value === 'true';
  }
})
booleanAttribute = true;
```

**Reply by dbwodlf3** [reply]
@deanhaleem umm.. actually, it is ambiguous but justinfagnanis saying, it is just a trivial.
nomrally boolean variables name is isEditable, isEditing or something something..
anyway, default value false is more natural.
my mind is changed..(default boolean value is changed to false..)

**Reply by deanhaleem** [reply]
Yep I agree `false` is more natural. In my case, I'm migrating a vue component to a lit element, and I don't want to make all the consumers change to needing to explicitly pass in `true`. So I was jw what a decent workaround could be 🤔

**Reply by AndrewJakubowicz** [reply]
There's a tracking issue to add documentation to lit.dev that boolean attributes must default to false. The issue includes some additional references: https://github.com/lit/lit.dev/issues/744
[The code for backwards compatibility](https://github.com/lit/lit/discussions/2959#discussioncomment-3202832) looks great and has the API of an [enumerated attribute](https://html.spec.whatwg.org/dev/common-microsyntaxes.html#keywords-and-enumerated-attributes). My only comment would be also handling invalid values for example: `<my-el boolean-attribute="invalid"></my-el>`

**Reply by deanhaleem** [reply]
Thanks appreciate the resources and feedback! 😄

**Comment by sorvell** [comment]
> So I was jw what a decent workaround could be 

How about binding the value to the property instead of the attribute. Here's an [example](https://lit.dev/playground/#gist=877c5148472fc129a2c0f8cfa623e562).

**Reply by deanhaleem** [reply]
So the issue for me is that I'm using the component in a Vue 2 app. So I only have `v-bind`. So ya, your sample doesn't work in Vue 2 template which is what we use 😕 . 

Specifically, it looks like Vue 2 [executes `removeAttribute`](https://github.com/vuejs/vue/blob/1d8b892fcd9edcef1feeed74ad586beabb9f6df2/src/platforms/web/runtime/modules/attrs.ts#L64) before any attribute is actually on the element, so the `attributeChangedCallback` is never invoked. And so Lit never knows to set the property to false. At least, I think that's what's happening. 

I keep emphasizing vue 2 since in vue 3 they do [something else](https://github.com/vuejs/core/blob/a95554d35c65e5bfd0bf9d1c5b908ae789345a6d/packages/runtime-dom/src/modules/props.ts#L97), so it's not a problem.
